composer/datasets/streaming/dataset.py

```diff
@@ -180,34 +182,45 @@ def _done_loading(self) -> None:
         with self._lock:
             self._are_all_shards_downloaded = True
 
-    def _download_thread(self, shards: List[int], part_min_id: int, part_max_id: int) -> None:
+    def _download_thread(self, part_shards: List[int], part_shards_to_download: List[int], part_min_id: int,
+                         part_max_id: int) -> None:
         """Background thread to download and assimilate missing shards.
 
         Args:
-            shards (List[int]): The shards remaining to be downloaded.
+            part_shards (List[int]): The shards to be used by this partition.
+            part_shards_to_download (List[int]): The shards to be downloaded by this partition's worker (subset of ``part_shards``).
             part_min_id (int): Minimum sample ID of this partition.
             part_max_id (int): Maximum sample ID of this partition.
         """
-        shards = list(shards)
         if self.shuffle:
-            np.random.shuffle(shards)
-        for shard in shards:
+            # Always process first shard first because other workers may be waiting on it
+            first = part_shards[0:1]
+            rest = part_shards[1:]
+            np.random.shuffle(rest)
```

Comment:
np.random.shuffle(part_shards[1:])

Reply:
Ah interesting...  I think this onlyworks if I convert to numpy array first, but will do that. Looks cleaner.